### DISCUSSION TITLE
onedrive business account throttling

### DISCUSSION THREAD

**Discussion by addohm** [abraunegg/onedrive#2482] [General]
Realizing this has been brought up before here, I've been unable to find what settings or where . 
 I recently started a business account.  I am trying to upload the contents of my backed up cloud drive and for over a week now I've been throttled non-stop.  Not a single upload (or download) is occuring. 

![image](https://github.com/abraunegg/onedrive/assets/25333107/0f495999-9032-40f0-b805-66532cd4f523)

 https://github.com/abraunegg/onedrive/issues/978

Now as one can expect of pretty much any tech company support, these guys don't know jack and when he heard "linux" all he could focus on was that.  I explained over and over that my personal account, for which I've had and used for over a decade, has _no_ issues.  I'm at my wit's end with this.  Can you tell me what setting and where?

**Comment by abraunegg** [comment]
@addohm 
Essentially it is the Microsoft OneDrive API and Service that is throttling you - so there is 100% zero setting or change that can be done in this client to resolve this situation.

However, you need to ask yourself - why is Microsoft doing this for your account when accessing the OneDrive API?

* How many active instances of the 'onedrive' client do you have running using the 'same' credentials?
  * On Windows and Linux and any mobile platform - identify what is going on
  * Microsoft OneDrive API is potentially thinking you are drying to perform a DDoS with those credentials
* Don't use the any OneDrive client with your credentials for 24-48 hours - ensure 'every' client you have running with your 'account' is shut down to allow Microsoft to see you have taken corrective action. This will help 'clear' the air on their end.

If you *still* are having an issue, as this *is* a Business Account - raise a support case with Microsoft and ask them *why* your account is being throttled for OneDrive API access. You get to raise a support case with Microsoft when using Business Accounts - so use that capability.

The last aspect is - you need to ensure that you are running the *latest* linux version - v2.4.25 - for your platform to ensure that there is not something else going on under the hood.

**Reply by shaun-xenon** [reply]
It is the API that is responsible for the throttling, you can read all about when throttling comes into play [https://learn.microsoft.com/en-us/sharepoint/dev/general-development/how-to-avoid-getting-throttled-or-blocked-in-sharepoint-online?WT.mc_id=365AdminCSH_globalsearch](here).

I've attached a screenshot of the limits and would like to know if we can configure the client to remain within these limits to avoid being repeatedly put to sleep, in my case, for 78 seconds each time. Thanks for your time and your efforts, they are very much valued and appreciated!
![Screenshot_20240223_131659](https://github.com/abraunegg/onedrive/assets/23053962/57340085-4fe8-444c-96f6-d3655fe8bef2)

**Reply by abraunegg** [reply]
API Limits are all *back end* controlled - there is *zero way* to control this client side.

**Comment by addohm** [comment]
I'll definitely log off all devices for the next couple of days.  

I'm definitely running 2.4.25.

I'm definitely running a business account.

Id say there's always two instances of onedrive running all the time.  My server, and my laptop.  Both under the same WAN IP.  \

At first I thought "Oh there's probably too frequent access" so I just left it.  I didn't expect this to still be a thing 1.5-2 weeks later.  My laptop and my server haven't done a damn thing.  They don't download, or upload.  What bothers me though is that my personal account works perfectly normal.  In fact, I just re-downloaded my entire personal account (850GB) without issue.

> @addohm Essentially it is the Microsoft OneDrive API and Service that is throttling you - so there is 100% zero setting or change that can be done in this client to resolve this situation.
> 
> However, you need to ask yourself - why is Microsoft doing this for your account when accessing the OneDrive API?
> 
> * How many active instances of the 'onedrive' client do you have running using the 'same' credentials?
>   
>   * On Windows and Linux and any mobile platform - identify what is going on
>   * Microsoft OneDrive API is potentially thinking you are drying to perform a DDoS with those credentials
> * Don't use the any OneDrive client with your credentials for 24-48 hours - ensure 'every' client you have running with your 'account' is shut down to allow Microsoft to see you have taken corrective action. This will help 'clear' the air on their end.
> 
> If you _still_ are having an issue, as this _is_ a Business Account - raise a support case with Microsoft and ask them _why_ your account is being throttled for OneDrive API access. You get to raise a support case with Microsoft when using Business Accounts - so use that capability.
> 
> The last aspect is - you need to ensure that you are running the _latest_ linux version - v2.4.25 - for your platform to ensure that there is not something else going on under the hood.

**Comment by abraunegg** [comment]
> At first I thought "Oh there's probably too frequent access" so I just left it. I didn't expect this to still be a thing 1.5-2 weeks later. My laptop and my server haven't done a damn thing. They don't download, or upload. What bothers me though is that my personal account works perfectly normal. In fact, I just re-downloaded my entire personal account (850GB) without issue.

In that case this will 100% be a Microsoft Account Issue. 

Raise a case with Microsoft.

**Comment by addohm** [comment]
"I don't support custom software" is all he keeps saying.

**Comment by abraunegg** [comment]
> "I don't support custom software" is all he keeps saying.

What you need to do is raise an issue saying:
* Microsoft OneDrive on Windows will not sync - constantly giving a 429 error
* Access of Micrsoft OneDrive via Microsoft Graph API gives the same result

If you tell them you are using Linux .. of course they will say not supported ... so tell them that accessing your data, using their software does not work.

**Comment by addohm** [comment]
The linux server syncs personal, but not business.  It allows upload and download from my laptop but not my server on the same network.

**Comment by addohm** [comment]
> 
I did this.  I've even showed them all my logs.

**Comment by addohm** [comment]
> If you tell them you are using Linux .. of course they will say not supported ... so tell them that accessing your data, using their software does not work.

Sadly that's not true lol.

I'll try detaching all devices for a few days to see what happens.  

I'll keep you up to date and thanks for trying to help.

**Comment by abraunegg** [comment]
> The linux server syncs personal, but not business. It allows upload and download from my laptop but not my server on the same network.

Hang on a second ... you said earlier that your Linux and Windows system is having the same problem, and both devices are using Business Account Credentials.

Can you please clarify more clearly what account type your Linux platform is using.

**Comment by addohm** [comment]
On my server (linux) it is using both personal and business until I can get this mess cleaned up.  My laptop is windows using the MS onedrive app.

**Comment by addohm** [comment]
Well, I fired it up after about 5 or 6 days of my account being "offline" or logged off from the one drive service.  Every device I have.  Still getting throttled.  No idea what to do but its clear that whatever this is it's very specific to the business account and Microsoft dolts don't know what the difference is.

**Reply by abraunegg** [reply]
@addohm 
Are you 100% sure that:

* The account that is being 'throttled' is actually a business account?

You mentioned that you have 2 instances working on the same system. Technically they 'should' be using different configurations + different access points.

I am happy to do a MS Teams Meeting if you are open to that to do some 'real-time' diagnosis to help understand what is going on.

**Comment by abraunegg** [comment]
@addohm 
As an FYI, with my 'Business Account' that I use for testing, I am triggering the following whilst attempting to upload 10K of randomly generated small files using `v2.5.0-alpha-0 GitHub version: v2.4.25-34-gbb1cf67` which runs 16 parallel threads for upload & download processes: 
```
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file249.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file259.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file260.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file254.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 40 seconds
ERROR: An error was returned from OneDrive and the resulting response is not a valid JSON object
ERROR: Increase logging verbosity to assist determining why.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file264.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file272.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file263.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file265.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file267.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file273.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file268.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file266.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file271.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file269.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file270.data ... done.
ERROR: An error was returned from OneDrive and the resulting response is not a valid JSON object
ERROR: Increase logging verbosity to assist determining why.
ERROR: An error was returned from OneDrive and the resulting response is not a valid JSON object
ERROR: Increase logging verbosity to assist determining why.
[M] Local file changed: random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file5.data
Scanning local filesystem 'random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN' for new data to upload ...
New items to upload to OneDrive: 226
Total New Data to Upload:        317 MB
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file275.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file287.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file286.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file281.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file278.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file277.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file282.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file285.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file279.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file289.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file284.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file274.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file283.data ... done.
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file288.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 5 seconds
Uploading new file random_10K_small_files/GpLvGBk6tno3R4PMs3fV5DoX4pw1v7VN/file280.data ... done.
```
Continues on for another 700 - 800 files then again:
```
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file233.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file226.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file235.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file227.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file242.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file252.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file253.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file247.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file251.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file244.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file249.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file240.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file246.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file248.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file250.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file241.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file254.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file243.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file245.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file255.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 64 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 64 seconds
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file266.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 60 seconds
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file267.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file260.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file269.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file268.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file270.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file259.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file271.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file257.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file264.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file258.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file261.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 62 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 61 seconds
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file265.data ... done.
```
Then ramps up dramatically:
```
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file338.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file343.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file351.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 209 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 326 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 305 seconds
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file360.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 324 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 144 seconds
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file361.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file364.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file356.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file353.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file358.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file362.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file359.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file357.data ... done.
Uploading new file random_10K_small_files/MiNikuBjehR8Z9EfzXo0K4HzOCNJ8MfU/file366.data ... done.
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 309 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 323 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 328 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 316 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 306 seconds
Thread sleeping due to 'HTTP request returned status code 429' - The request has been throttled
Sleeping for 306 seconds
```
Each thread eventually finishes and the uploads are all done - a few failed file in-between, but this is what I am doing to chase down issues with the v2.5.0-alpha-0 codebase at the moment.

**Comment by abraunegg** [comment]
@addohm 
At the moment I would suggest for you shut down your Business Docker container and try v2.4.25 as a stand alone application (not inside a Docker container) for your Business Account and see what happens - do you get the same throttling?

If there is still an issue, then it is 100% with your account that Microsoft need to look at. 

If you dont get the throttling, it is potentially something configuration wise or Docker deployment wise that has been done. I would suggest redeploy your 'Business' Docker container - re-configure from scratch, double check your steps.

**Comment by addohm** [comment]
Im still having this issue but missed one of your suggestions.  I half-way gave up on it to be honest and just started overusing the piss out of the personal account that I had intended on closing.  I don't think I ever tried this outside of docker.  Going to try this tonight.  

<img width="638" alt="Screenshot 2024-01-17 183702" src="https://github.com/abraunegg/onedrive/assets/25333107/81f46abb-7f31-4c29-a42d-1bcfba7bf0cd">

**Comment by addohm** [comment]
So far it seems to work outside of docker.
<img width="366" alt="image" src="https://github.com/abraunegg/onedrive/assets/25333107/a044f776-fa17-4203-8819-c8e7170281b5">
Note: this is not on the same machine.  I fired up a Pi5 for this.  @abraunegg Does that offer for a teams effort to troubleshoot this still stand?

**Comment by addohm** [comment]
Heres something thats more interesting.  
<img width="538" alt="image" src="https://github.com/abraunegg/onedrive/assets/25333107/db9c11ae-d5a5-4138-8ce1-e73d188eedcf">
Left and right, both the same docker container.  On the left the synchronize is ran automatically at startup.  On the right, I ran `onedrive --synchronize --verbose` manually.  I've left a message for the docker app maintainer.  https://forums.unraid.net/topic/133978-support-bversluis-onedrive/?do=findComment&comment=1355221

Based on [this old response to other issues](https://forums.unraid.net/topic/133978-support-bversluis-onedrive/?do=findComment&comment=1273185) I think it best if I create my own app package.

**Comment by abraunegg** [comment]
@addohm 
> Does that offer for a teams effort to troubleshoot this still stand?

So you are using Docker on some 'unraid' platform ...... using a Docker image or system that is not built by me.

My suggestions to you are:
* Seek support from 'unraid'
* The BVersluis image is running `onedrive --synchronize` ... which is *wrong* for a Docker image
  * The Docker image should be using `onedrive --monitor` at startup - so if it is doing something else .. then thats a modification I am not responsible for
* Use the correct Docker image to sync your data - please read:
  * https://github.com/abraunegg/onedrive/blob/master/docs/Docker.md
  * https://github.com/abraunegg/onedrive/blob/master/docs/Podman.md

Please seek support from 'Unraid' for your issue as they are providing the Operating System on which you are running. You bought it from them - thus you need to seek support from the 'unraid' folk.

Their system / solution / method for Docker networking are contributing factors to your issue. Given that you can run the client manually, on the same network, but on a different machine without issue - points the issue to your 'unraid' box.

Your alternative is to not use 'unraid' .. build yourself a system using Debian 12 or CentOS 8 Stream, use ZFS for the data storage for data you care about, run Docker or Podman in a supported way.

**Reply by addohm** [reply]
Nah, its still your image.  They just create an "app" layer over it that basically creates the docker compose file.  It literally does nothing except add convenience to the people that aren't good with docker.  

You can see what it does here: https://github.com/addohm/unraid-templates/blob/main/Onedrive/Onedrive.xml

Anyway, for now it seems to be running correctly outside of some other issues I'm trying to resolve (also here in discussion).
